### PR TITLE
Fix zbsniff to correctly extract Network Key.

### PR DESCRIPTION
**Pull request by SteveJM**
Refactor to correctly decrypt network key - base change.
Update to compute nonce from cached iEEE Addresses.
Add improved command line arguments.
Bring to production quality.
Tested against sample captures.

### CODE REVIEW

File: tools/zbdsniff [hunk 1]
```diff
@@ -27,98 +52,113 @@ def getnetworkkey(packet):
...
+(options, args) = parser.parse_args()
 
+if (not options.filename and not options.directory):
+    print("A packet capture file or directory must be specified")
+    sys.exit(1)
 
-if len(sys.argv) == 1:
-    usage()
-    sys.exit(0)
+if (not options.transportKey):
```
Comment (rmspeers) [inline tools/zbdsniff]: Often users will want to default to the HA1.2 spec'ed key, but I think it's reasonable for them to be required to provide this -- it shows they think about what is going on!

File: tools/zbdsniff [hunk 2]
```diff
@@ -27,98 +52,113 @@ def getnetworkkey(packet):
...
 
 files = []
-filecount = 0
+if options.filename:
+    files.append(options.filename)
 
-while len(sys.argv) > 1:
-    files.append(sys.argv.pop(1))
+if options.directory:
+    files += glob.glob(options.directory + os.sep + "*.pcap")
```
Comment (rmspeers) [inline tools/zbdsniff]: prefer os.join() but this is the equivalent and is platform safe so will be OK

File: tools/zbdsniff (hunk 1 above)
Comment (SteveJM) [reply tools/zbdsniff]: I was concerned about putting any "default" key in that might lead to the repo being flagged or reported by the Alliance.

File: tools/zbdsniff (hunk 2 above)
Comment (SteveJM) [reply tools/zbdsniff]: Hmmm, I'll go for os.path.join() as that delegates the responsibility entirely.

**Comment by SteveJM** [comment]
I've now updated to use Scapy.

**Comment by SteveJM** [comment]
Bump.

File: tools/zbdsniff [hunk 3]
```diff
@@ -1,149 +1,110 @@
...
+                    else:
+                        print("[-] There was no ext_src mapping for: ", str(nwk.source))
+                else:
+                    print("[-] Unexpected key_type", str(sec.key_type))
+
+
+if __name__ == '__main__':
+    # Define the command line options.
+    parser = OptionParser(description="zbdsniff: Decode plaintext Zigbee Nwtwork key from a " +
+        "capture file. Will process libpcap or Daintree SNA capture files. Original concept: " +
```
Comment (rmspeers) [inline tools/zbdsniff]: I think we'll now only support libpcap than also Daintree. Given the standard branching strategy, we would hold this for a major version release, 3.0. However, given the small usage of Daintree SNA and the other benefits of this new PR being much cleaner/better tested/etc -- and our disclaimers on the KillerBee project being active/etc -- I think this is fine to put into develop now.

File: tools/zbdsniff [hunk 4]
```diff
@@ -1,149 +1,110 @@
...
+
+    if options.directory:
+        files += glob.glob(os.path.join(options.directory, "*.pcap"))
+
+    filecount = 0
+
+    for file in files:
+        print "Processing %s"%file
+        if not os.path.exists(str(file)):
+            print >>sys.stderr, "ERROR: Input file \"%s\" does not exist." % file
```
Comment (rmspeers) [inline tools/zbdsniff]: develop doesn't require py3 yet, but we should do py3 for new development to support a later transition. However, I will handle this in the py3 branch.

**Review by rmspeers — APPROVED** [review]

**Comment by rmspeers** [comment]
@SteveJM  I have put this in the Python 2.7.1 branch with some tweaks for error handling of Daintree, and for python3 forwards compatibility in zbdsniff.

Is there a PCAP file you could contribute to `samples/` for use in examples as well as future unit testing?

**Comment by Sandb0xED** [comment]
Hi there, I'm trying to understand through these threads how to fix the "error of the input file not able to be loaded. Is it a PCAP file? Daintree support was removed in Killerbee 2.7.1". It just cannot recognize the pcap file (which was created by zbwireshark). I tried running pcapfix or editcap on it, but it still does not accept the file. Any ideas/recommendations?

**Comment by riverloopsec** [comment]
What command line are you using?

Also, can you share the PCAP file so we can try to reproduce?

**Pull request merged**